resourcemanager/pool/spool/spool.go

```diff
@@ -0,0 +1,177 @@
+	capacity           int32
+	running            int32
+	isStop             atomic.Bool
+	concurrencyMetrics prometheus.Gauge
+	taskManager        pooltask.TaskManager[any, any, any, any, pooltask.NilContext]
+	pool.BasePool
+}
+
+// NewPool create a single goroutine pool.
+func NewPool(name string, size int32, component util.Component, options ...Option) (*Pool, error) {
+	opts := loadOptions(options...)
+	result := &Pool{
+		options:            opts,
+		concurrencyMetrics: metrics.PoolConcurrencyCounter.WithLabelValues(name),
+		taskManager:        pooltask.NewTaskManager[any, any, any, any, pooltask.NilContext](size), // TODO: this general type
+	}
+	if size == 0 {
+		return nil, pool.ErrPoolParamsInvalid
+	}
+	result.SetName(name)
+	result.capacity = size
+	result.concurrencyMetrics.Set(float64(size))
+	err := resourcemanager.InstanceResourceManager.Register(result, name, component)
+	if err != nil {
+		return nil, err
+	}
+	return result, nil
+}
+
+// Tune changes the pool size.
+func (p *Pool) Tune(size int) {
+	if size == 0 {
+		return
+	}
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	p.SetLastTuneTs(time.Now())
+	p.capacity = int32(size)
+	p.concurrencyMetrics.Set(float64(size))
+}
+
+// Run runs a function in the pool.
+func (p *Pool) Run(fn func()) error {
+	if p.isStop.Load() {
+		return pool.ErrPoolClosed
+	}
+	_, run := p.checkAndAddRunning(1)
+	if !run {
+		return pool.ErrPoolOverload
+	}
+	p.taskManager.RegisterTask(p.GenTaskID(), 1)
+	p.run(fn)
+	return nil
+}
+
+// Cap returns the capacity of the pool.
+func (p *Pool) Cap() int {
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+	return int(p.capacity)
+}
+
+// Running returns the number of running goroutines.
+func (p *Pool) Running() int {
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+	return int(p.running)
+}
+
+func (p *Pool) run(fn func()) {
+	p.wg.Add(1)
+	go func() {
+		defer func() {
+			if r := recover(); r != nil {
+				logutil.BgLogger().Error("recover panic", zap.Any("recover", r))
+			}
+			p.wg.Done()
+			p.mu.Lock()
+			p.running = p.running - 1
+			p.mu.Unlock()
+		}()
+		fn()
+	}()
+}
+
+// RunWithConcurrency runs a function in the pool with concurrency.
+func (p *Pool) RunWithConcurrency(fns chan func(), concurrency uint32) error {
+	if p.isStop.Load() {
+		return pool.ErrPoolClosed
+	}
+	conc, run := p.checkAndAddRunning(concurrency)
+	if !run {
+		return pool.ErrPoolOverload
+	}
+	// TODO: taskManager need to refactor
+	p.taskManager.RegisterTask(p.GenTaskID(), conc)
+	for n := int32(0); n < conc; n++ {
+		p.run(func() {
+			for fn := range fns {
+				fn()
+			}
+		})
+	}
+	return nil
+}
+
+// checkAndAddRunning is to check if a task can run. If can, add the running number.
+func (p *Pool) checkAndAddRunning(concurrency uint32) (conc int32, run bool) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	for {
+		value, run := p.checkAndAddRunningInternal(int32(concurrency))
+		if run {
+			return value, run
+		}
+		if !p.options.Blocking {
+			return 0, false
+		}
+		time.Sleep(5 * time.Millisecond)
```

Comment:
If one of the request is blocked, the user cannot even query `pool.Running()` or `pool.Cap()`. Is it expected?

Reply:
No, so I recovered the atomic of running and cap.

Why will not the atomic of running and cap affect the tidb?

1、the add of running is under the control of mutex, so all the request will wait for the running value decrease when to task finish.

2.Cap isn't changed in the mutex, it can be changed in the ```Tune```. it will happened ```Tune``` is changed after the ```Task``` running. But the Task manager will control the running to meet the capacity.

Reply:
Done